api/v1alpha1/openstackrouter_types.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// OpenStackRouterSpec defines the desired state of OpenStackRouter
+type OpenStackRouterSpec struct {
+	// Cloud is the OpenStackCloud hosting this resource
+	Cloud string `json:"cloud"`
+
+	// ID is the OpenStack UUID of the resource. If left empty, the
+	// controller will create a new resource and populate this field. If
+	// manually populated, the controller will adopt the corresponding
+	// resource.
+	ID string `json:"id,omitempty"`
+
+	// Name of the OpenStack resource.
+	Name string `json:"name,omitempty"`
+
+	Description string `json:"description,omitempty"`
+
+	ExternalGatewayInfo OpenStackRouterExternalGatewayInfo `json:"externalGatewayInfo,omitempty"`
+
+	// AdminStateUp is the administrative state of the router.
+	AdminStateUp *bool `json:"adminStateUp,omitempty"`
+
+	// Distributed is whether router is distributed or not.
+	Distributed *bool `json:"distributed,omitempty"`
+
+	// Availability zone hints groups router nodes.
+	// Used to make router resources highly available.
+	AvailabilityZoneHints []string `json:"availabilityZoneHints,omitempty"`
+
+	// TenantID is the project owner of the router. Only admin users can
+	// specify a project identifier other than its own.
+	TenantID string `json:"tenantID,omitempty"`
+
+	// ProjectID is the project owner of the router.
+	ProjectID string `json:"projectID,omitempty"`
+
+	// All the subnets that will be added as router interfaces
+	Subnets []string `json:"subnets,omitempty"`
```

Comment:
I've been thinking a lot about this one. I agree that we must provide a way to add interfaces right from this spec... the problem now is _how_.

As far as I know, there are two ways to add an interface to a router:
* call `add_router_interface` with a `subnet_id` (which is what you do here); OR
* create a port and call `add_router_interface` with its `port_id`.

I think we should only provide a way to adding ports, and not subnets, to routers. The reason for this is: when you pass a subnet, Neutron automatically creates a port, which is not represented anywhere in our CRs. Instead if you create a port (for which we will get a CRD soon) and only then add it to a router, then everything is nicely accounted for in our YAMLs.

By adding `Ports` you enable users to add a specific IP to their router, while with `Subnets` you only allow ports bound to the subnet's gateway IP. In that respect, passing a `port_id to `add_router_interface` is a strict superset of passing a `subnet_id`.

---

My proposal: remove `Subnets` and add `Port` instead... which accepts a port ID until we add `OpenStackPort` to the CRDs.

---

What do you think?

Reply:
The current implementation was done taking in account the fact that we were developing the Router before the Port, so only one way of adding interfaces could exist (via Subnet).

Having that in mind I think is that it is much more intuitive to add Subnets directly than having to create a Port. The port that Neutron creates is included in the interfacesinfo. What we could do is to create Port CRs when the information from Neutron comes to have them represented in our CRs.

Of course, providing a Port should also be supported when it is possible, but I think the use case of providing a Subnet would be principal one.

Maybe it would be a good approach to merge this with Subnet support and then contribute a new PR adding the support to Ports when that is included (at least, that was my intention).